server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -588,6 +622,223 @@ private Scheduler.Cancellable fetchClusterFormationInfo(
         }, new TimeValue(10, TimeUnit.SECONDS), ThreadPool.Names.SAME);
     }
 
+    private void beginPollingRemoteStableMasterHealthIndicatorService(Collection<DiscoveryNode> masterEligibleNodes) {
+        synchronized (remoteDiagnosticsMutex) {
+            if (remoteCoordinationDiagnosticsCancelled.get()) { // Don't start a 2nd one if the last hasn't been cancelled
```

Comment:
I'll add more comments. `remoteCoordinationDiagnosticsCancelled` is a reference to the current/most recent poll. It is kept as an object-level variable so that `cancelPollingRemoteStableMasterHealthIndicatorService()` can cancel it. `isCancelled` is a reference to the one for a specific poll, which is why it's kept on stack. So if we get to a point where there are two polls running at once, the older one will check its copy of `isCancelled` and see that `cancelPollingRemoteStableMasterHealthIndicatorService()` has cancelled it (even if the current value of `remoteCoordinationDiagnosticsCancelled` is `true`.

Reply:
> The synchronized block can be avoided here by doing a `compareAndSwap` on the `AtomicBolean` - I renamed `remoteCoordinationDiagnosticsCancelled` to `isRemoteFetchRunning` but some better naming is needed here :)

The reason for the mutex is to protect the assignment of 3 variables atomically -- `remoteStableMasterHealthIndicatorTask`, `remoteCoordinationDiagnosisResult`, and `remoteCoordinationDiagnosticsCancelled`, all while knowing that `remoteCoordinationDiagnosticsCancelled` is `true`. I don't think your suggestion covers that, does it?

Reply:
These assignments happen inside the proposed CAS operation 
ie. 
```
 if (isRemoteFetchRunning.compareAndSwap(false, true)) {
```
The **one thread** that gets to flip this boolean will be granted access to the contents of this if-block. 
I believe it would be covered.